File: backend/app/models/sofa_loan_entry.py
from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, Numeric, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class SofaLoanEntry(Base):
    __tablename__ = "sofa_loan_entries"
    __table_args__ = (UniqueConstraint("sofa_loan_id", "entry_month", name="uq_loan_month"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    sofa_loan_id: Mapped[int] = mapped_column(ForeignKey("sofa_loans.id"), nullable=False)
    entry_month: Mapped[date] = mapped_column(Date, nullable=False)
    disbursed: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    repayment: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    interest_collected: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    loan: Mapped["SofaLoan"] = relationship("SofaLoan", back_populates="entries")  # noqa: F821

File: frontend/lib/models/sofa_loan_entry.dart
class SofaLoanEntry {
  final int id;
  final int sofaLoanId;
  final String entryMonth;
  final double disbursed;
  final double repayment;
  final double interestCollected;
  final DateTime createdAt;
  final DateTime updatedAt;

  const SofaLoanEntry({
    required this.id,
    required this.sofaLoanId,
    required this.entryMonth,
    required this.disbursed,
    required this.repayment,
    required this.interestCollected,
    required this.createdAt,
    required this.updatedAt,
  });

  factory SofaLoanEntry.fromJson(Map<String, dynamic> json) {
    final now = DateTime.now();
    return SofaLoanEntry(
      id: json['id'] as int,
      sofaLoanId: json['sofa_loan_id'] as int,
      entryMonth: json['entry_month'] as String,
      disbursed: (json['disbursed'] as num).toDouble(),
      repayment: (json['repayment'] as num).toDouble(),
      interestCollected: (json['interest_collected'] as num).toDouble(),
      createdAt: DateTime.tryParse(json['created_at'] as String? ?? '') ?? now,
      updatedAt: DateTime.tryParse(json['updated_at'] as String? ?? '') ?? now,
    );
  }
}
